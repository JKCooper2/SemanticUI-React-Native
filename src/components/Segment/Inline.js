// @flow
import React, { PureComponent } from 'react';
import { View, Text } from 'react-native';
import _ from 'lodash';

import type { InlineProps } from '~/types/components/Segment';

class Inline extends PureComponent<InlineProps> {
    render() {
        let { children } = this.props;

        let childComp = React.Children.map(children, child => (_.isString(child)
            ? <Text>{child}</Text>
            : child
        ));

        return (
            <View style={{ flexDirection: 'row' }}>
                {childComp}
            </View>
        );
    }
}

export default Inline;