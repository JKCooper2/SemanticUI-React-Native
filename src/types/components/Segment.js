// @flow
import type { Node } from 'react';
import type { ColorsType } from '~/types/styles';

export type SegmentProps = {
    +children: Node;
    +content?: string;
    +placeholder: boolean;
    +padded: boolean | "very";
    +compact: boolean;
    +color?: ColorsType;
    +inverted: boolean;
    +secondary: boolean;
    +tertiary: boolean;
    +circular: boolean;
    +style?: Object;
    +basic: boolean;
}

export type InlineProps = {
    +children: Node;
}