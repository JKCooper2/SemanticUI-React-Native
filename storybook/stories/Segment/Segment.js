// @flow
import React from 'react';
import { View, ScrollView } from 'react-native';
import { storiesOf } from '@storybook/react-native';

import Header from '~/components/Header';
import Button from '~/components/Button';
import Segment from '~/components/Segment';

const stories = storiesOf('Segment', module);

stories.add('Standard', () => (
    <ScrollView style={{ padding: 10, backgroundColor: 'white' }}>
        <Segment>Here is a segment</Segment>
        <Segment basic>Here is a basic segment</Segment>
        <Segment padded>Here is a padded segment</Segment>
        <Segment padded="very">Here is a very padded segment</Segment>
        <Segment compact>Here is a compact segment</Segment>
        <Segment placeholder>Here is a placeholder segment</Segment>
        <Segment placeholder>
            Here is an inline segment
            <Segment.Inline>
                <Button primary>Clear Query</Button>
                <Button>Add Document</Button>
            </Segment.Inline>
        </Segment>
        <Segment color="red">Here is a colored segment</Segment>
        <Segment color="blue" inverted>Here is an inverted colored segment</Segment>
        <Segment secondary>Here is a secondary segment</Segment>
        <Segment tertiary>Here is a tertiary segment</Segment>
        <Segment inverted>Here is an inverted segment</Segment>
        <Segment inverted secondary>Here is an inverted secondary segment</Segment>
        <Segment inverted tertiary>Here is an inverted tertiary segment</Segment>
        <View style={{ flexDirection: 'row' }}>
            <Segment circular style={{ width: 175, height: 175 }}>
                <Header as="h4">
                    Sale!
                    <Header.SubHeader as="h6">$10.99</Header.SubHeader>
                </Header>
            </Segment>
            <Segment circular inverted style={{ width: 175, height: 175 }}>
                <Header as="h4" inverted>
                    Buy Now
                    <Header.SubHeader as="h6">$10.99</Header.SubHeader>
                </Header>
            </Segment>
        </View>
    </ScrollView>
));

export default stories;
