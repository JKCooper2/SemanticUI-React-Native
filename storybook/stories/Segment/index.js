// @flow
import Segment from './Segment';

export {
    Segment as default
};