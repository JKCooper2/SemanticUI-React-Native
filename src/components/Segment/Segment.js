// @flow
import React, { PureComponent } from 'react';
import { View, Text } from 'react-native';
import _ from 'lodash';

import { spacing, colors } from '~/styles';

import type { SegmentProps } from '~/types/components/Segment';
import type { ColorsType } from '~/types/styles';

import Inline from './Inline';

const getPadding = (padding: boolean | "very"): number => {
    if (padding === 'very') {
        return spacing.padding.huge;
    }

    if (padding === true) {
        return spacing.padding.large;
    }

    return spacing.padding.medium;
};

const getBackgroundColor = (placeholder: boolean, color: ?ColorsType, inverted: boolean, secondary: boolean, tertiary: boolean): string => {
    if (inverted) {
        let backgroundColor = _.get(colors.colors, color, colors.colors.black);

        if (secondary) {
            return `${backgroundColor}AA`;
        }

        if (tertiary) {
            return `${backgroundColor}44`;
        }

        return backgroundColor;
    }

    if (secondary || placeholder) {
        return colors.colors.lightestGrey;
    }

    if (tertiary) {
        return colors.colors.lightGrey;
    }

    return colors.colors.white;
};

const getFontColor = (inverted: boolean, secondary: boolean, tertiary: boolean): ColorsType => {
    if (inverted && (secondary || tertiary)) {
        return colors.colors.lightGrey;
    }

    if (inverted) {
        return colors.colors.white;
    }

    if (secondary || tertiary) {
        return colors.colors.darkGrey;
    }

    return colors.colors.black;
};

const getBorderColor = (backgroundColor: string): string => {
    if (backgroundColor === colors.colors.white) {
        return colors.colors.lightGrey;
    }

    return backgroundColor;
};

// TODO: Figure out box-shadow
class Segment extends PureComponent<SegmentProps> {
    static Inline = Inline;

    static defaultProps = {
        placeholder: false,
        padded: false,
        compact: false,
        inverted: false,
        secondary: false,
        tertiary: false,
        style: {},
        circular: false,
        basic: false
    }

    render() {
        let { children, placeholder, padded, compact, color, inverted, secondary, tertiary, style, circular, basic } = this.props;

        let fontColor = getFontColor(inverted, secondary, tertiary);

        let childComp = React.Children.map(children, child => (_.isString(child)
            ? <Text style={{ color: fontColor }}>{child}</Text>
            : child
        ));

        let backgroundColor = getBackgroundColor(placeholder, color, inverted, secondary, tertiary);
        let borderColor = getBorderColor(backgroundColor);

        let borderTopWidth = 1;
        if (color && !inverted) {
            borderTopWidth = 2;
        }
        if (basic) {
            borderTopWidth = 0;
        }

        return (
            <View style={{
                alignSelf: compact ? 'flex-start' : 'stretch',
                alignItems: (placeholder || circular) ? 'center' : 'flex-start',
                justifyContent: circular ? 'center' : 'flex-start',
                backgroundColor,
                padding: getPadding(padded),
                borderColor,
                borderStyle: 'solid',
                borderWidth: basic ? 0 : 1,
                borderTopWidth,
                borderTopColor: _.get(colors.colors, color, colors.colors.lightGrey),
                borderRadius: circular ? 1000 : spacing.border.medium,
                ...style
            }}>
                {childComp}
            </View>
        );
    }
}

export default Segment;