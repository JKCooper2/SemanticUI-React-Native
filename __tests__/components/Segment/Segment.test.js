// @flow
import React from 'react';
import { shallow } from 'enzyme';

import Segment from '~/components/Segment';

test('renders correctly', () => {
    const segment = shallow(<Segment>Segment</Segment>);
    expect(segment).toMatchSnapshot();
});
