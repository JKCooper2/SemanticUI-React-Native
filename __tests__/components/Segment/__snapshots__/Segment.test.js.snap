// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Object {
      "alignItems": "flex-start",
      "alignSelf": "stretch",
      "backgroundColor": "#fff",
      "borderColor": "#e8e8e8",
      "borderRadius": 5,
      "borderStyle": "solid",
      "borderTopColor": "#e8e8e8",
      "borderTopWidth": 1,
      "borderWidth": 1,
      "justifyContent": "flex-start",
      "padding": 20,
    }
  }
>
  <Component
    key=".0"
    style={
      Object {
        "color": "#1b1c1d",
      }
    }
  >
    Segment
  </Component>
</Component>
`;
